src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs

```diff
@@ -188,60 +331,70 @@ SyntaxToken getCloseQuote(Range openQuoteRange)
+                    kind is Lexer.InterpolatedStringKind.Normal or Lexer.InterpolatedStringKind.Verbatim
+                        ? SyntaxKind.CloseBraceToken
+                        : SyntaxKind.RawInterpolationCloseToken,
+                    text[interpolation.CloseBraceRange],
+                    trailing: null);
             }
         }
 
         /// <summary>
         /// Interpret the given raw text from source as an InterpolatedStringTextToken.
         /// </summary>
         /// <param name="text">The text for the full string literal, including the quotes and contents</param>
-        /// <param name="kind">The kind of the interpolation</param>
-        private SyntaxToken MakeInterpolatedStringTextToken(string text, Lexer.InterpolatedStringKind kind)
+        /// <param name="kind">The kind of the interpolated string we were processing</param>
+        private SyntaxToken MakeInterpolatedStringTextToken(Lexer.InterpolatedStringKind kind, string text)
         {
+            // with a raw string, we don't do any interpretation of the content.  Note: removal of indentation is
+            // handled already in splitContent
+            if (kind is Lexer.InterpolatedStringKind.SingleLineRaw or Lexer.InterpolatedStringKind.MultiLineRaw)
+                return SyntaxFactory.Literal(leading: null, text, SyntaxKind.InterpolatedStringTextToken, text, trailing: null);
+
+            Debug.Assert(kind is Lexer.InterpolatedStringKind.Normal or Lexer.InterpolatedStringKind.Verbatim);
+
+            // For a normal/verbatim piece of content, process the inner content as if it was in a corresponding
+            // *non*-interpolated string to get the correct meaning of all the escapes/diagnostics within.
             var prefix = kind is Lexer.InterpolatedStringKind.Verbatim ? "@\"" : "\"";
             var fakeString = prefix + text + "\"";
-            using var tempLexer = new Lexer(Text.SourceText.From(fakeString), this.Options, allowPreprocessorDirectives: false);
+            using var tempLexer = new Lexer(SourceText.From(fakeString), this.Options, allowPreprocessorDirectives: false);
 
             var mode = LexerMode.Syntax;
             var token = tempLexer.Lex(ref mode);
             Debug.Assert(token.Kind == SyntaxKind.StringLiteralToken);
             var result = SyntaxFactory.Literal(leading: null, text, SyntaxKind.InterpolatedStringTextToken, token.ValueText, trailing: null);
             if (token.ContainsDiagnostics)
-            {
                 result = result.WithDiagnosticsGreen(MoveDiagnostics(token.GetDiagnostics(), -prefix.Length));
-            }
 
             return result;
         }
 
         private static DiagnosticInfo[] MoveDiagnostics(DiagnosticInfo[] infos, int offset)
         {
-            var builder = ArrayBuilder<DiagnosticInfo>.GetInstance();
+            if (offset == 0)
+                return infos;
+
+            var builder = ArrayBuilder<DiagnosticInfo>.GetInstance(infos.Length);
```

Comment:
nit: could assert positive length (ie. we don't need a quick exit path when no diagnostics, as this method simply won't be called)

Not related to this PR: Just curious about `var sd = info as SyntaxDiagnosticInfo;` below. Can the cast ever fail?

Reply:
done!

Reply:
yeah... that cast failing would be highly suspect.  i'll make a normal cast.